### PR TITLE
Study setting for deleting unsentable emails

### PR DESCRIPTION
**Pull request by yufengwng**
## Context

In `dev` environment, we're getting alerts of this nature:
```
Could not create message for queued event 25950 Cannot send email to participant LYAC0LZT4L4JV039JPRP
with no auth0 account and no proxies in study testboston
```

This is happening because we have feature where study admin creates a new participant who does not have an auth0 email and does not have proxies. Therefore, there is no place to send emails to. For other studies, this is considered an error situation, thus the error log. But in testboston study this is allowed. To keep the behavior the same for other studies and allowing this at the same time for testboston, we're adding a new study setting to control this: `shouldDeleteUnsentableEmails`. When this flag is true, queued email events that can't be sent to an email address will be deleted.

## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [x] I have considered error handling and alerts, and added `L-*` labels as needed.
- [x] I have considered security and privacy, and added `I-*` labels as needed
- [x] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.

## FUD Score

_Overall, how are you feeling about these changes?_

- [x] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [x] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [x] I have written automated positive tests
- [ ] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/src/main/resources/db-changes/schema/should-delete-unsentable-emails.xml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+                   http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.5.xsd">
+
+    <changeSet author="yufeng" id="20200629-should-delete-unsentable-emails">
+        <addColumn tableName="study_settings">
+            <column name="should_delete_unsentable_emails" type="boolean" defaultValueBoolean="false">
+                <constraints nullable="false"/>
```
Comment (MocanaAtBroad) [inline pepper-apis/src/main/resources/db-changes/schema/should-delete-unsentable-emails.xml]: _unsen**d**able_

File: pepper-apis/src/main/java/org/broadinstitute/ddp/model/study/StudySettings.java [hunk 2]
```diff
@@ -9,17 +9,20 @@
...
+    private boolean shouldDeleteUnsentableEmails;
 
     @JdbiConstructor
     public StudySettings(
             @ColumnName("umbrella_study_id") long studyId,
             @ColumnName("invite_error_template_id") Long inviteErrorTemplateId,
             @ColumnName("analytics_enabled") boolean analyticsEnabled,
-            @ColumnName("analytics_token") String analyticsToken) {
+            @ColumnName("analytics_token") String analyticsToken,
+            @ColumnName("should_delete_unsentable_emails") boolean shouldDeleteUnsentableEmails) {
```
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/model/study/StudySettings.java]: I think you mean shouldDeleteUnsenda**b**leEmails . Applies to all usages of `unsentable` in code

File: pepper-apis/src/main/java/org/broadinstitute/ddp/Housekeeping.java [hunk 3]
```diff
@@ -367,8 +367,15 @@ public static void main(String[] args) {
                                         try {
                                             message = messageBuilder.createMessage(ddpMessageId, pendingEvent, apisHandle);
                                         } catch (MessageBuilderException e) {
-                                            LOG.error("Could not create message for queued event "
-                                                    + pendingEvent.getQueuedEventId(), e);
+                                            if (e.shouldDeleteEvent()) {
+                                                LOG.warn("Unable to create message for event with queued_event_id={}, proceeding to delete",
+                                                        pendingEvent.getQueuedEventId(), e);
+                                                queuedEventDao.deleteAllByQueuedEventId(pendingEvent.getQueuedEventId());
+                                                return; // Exit out of transaction wrapper and move on to next event.
```
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/Housekeeping.java:381]: Design-wise does not look right to me to have exception itself dictate to code catching it what to do (shouldDeleteEvent). A suggestion: maybe there is an "AddressNotAvailableException" (maybe subclass of MessageBuilder) and then the handler, based on the Study configuration setting you created in the PR, can decide on whether to create the event or not.

File: pepper-apis/src/main/resources/db-changes/schema/should-delete-unsentable-emails.xml (hunk 1 above)
Comment (yufengwng) [reply pepper-apis/src/main/resources/db-changes/schema/should-delete-unsentable-emails.xml]: Thanks!

File: pepper-apis/src/main/java/org/broadinstitute/ddp/Housekeeping.java (hunk 3 above)
Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/Housekeeping.java:381]: Good point! I have moved the decision higher up in the caller now.

**Review by MocanaAtBroad — APPROVED** [review]

**Pull request merged**